pex/resolve/locker.py

```diff
@@ -315,129 +335,75 @@ def analyze(self, line):
+                    source_archive_path = os.path.join(self._download_dir, selected_path)
+                    if not os.path.isfile(source_archive_path):
+                        raise AnalyzeError(
+                            "Failed to lock {artifact}. Could not obtain its content for "
+                            "analysis.".format(artifact=artifact_url)
                         )
-                    elif self._source_distribution_partial_artifact:
-                        partial_artifact = self._source_distribution_partial_artifact
-                        self._source_distribution_partial_artifact = None
-
-                        url_info = urlparse.urlparse(partial_artifact.url)
-                        source_distribution_path = unquote(url_info.path)
-                        if "file" != url_info.scheme:
-                            source_distribution_path = os.path.join(
-                                self._download_dir,
-                                os.path.basename(source_distribution_path),
-                            )
-                        if not os.path.exists(source_distribution_path):
-                            raise AnalyzeError(
-                                "Failed to lock {artifact}. Could not obtain its content for "
-                                "analysis.".format(artifact=partial_artifact)
-                            )
-
-                        digest = Sha256()
-                        if os.path.isdir(source_distribution_path):
-                            digest_local_project(
-                                directory=source_distribution_path,
-                                digest=digest,
-                                pip_version=self._pip_version,
-                                resolver=self._resolver,
-                            )
-                        else:
-                            hashing.file_hash(source_distribution_path, digest)
-                        fingerprint = digest.hexdigest()  # type: hashing.Fingerprint
-
-                        self._resolved_requirements.append(
-                            ResolvedRequirement(
-                                requirement=requirement,
-                                pin=pin,
-                                artifact=attr.evolve(
-                                    partial_artifact,
-                                    fingerprint=Fingerprint(
-                                        algorithm=fingerprint.algorithm, hash=fingerprint
-                                    ),
-                                    verified=True,
-                                ),
-                            )
+                    digest = Sha256()
+                    hashing.file_hash(source_archive_path, digest)
+                    source_fingerprint = Fingerprint.from_digest(digest)
+                    self._selected_path_to_pin[selected_path] = build_result.pin
+                elif "file" == artifact_url.scheme:
```

Comment:
Hard for me to tell by eyeballing, but are we certain we will never encounter a schemeless URL (i.e., just a file path) in the parsed output, that will eventually get passed here? 

In other words, should also check for a None scheme and treat it like a file: ?

Reply:
Yeah, certain. ITs cover that case across Pex's closed universe of Pips. Pip records all of `.`, `$PWD` and `file://$PWD`  as `file://$PWD` in the logs. The only anomaly is in its `Saved ...` log lines (which are used by the lock log analyzer to filter out backtracks). In those lines local file paths that are directories never appear. This one anomaly carries a fair bit of supporting code and tests of its own.